### PR TITLE
Rework `itertools`, add col-to-row and row-to-col functions.

### PR DESCRIPTION
**Pull request by jaheba**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** BREAKING, new feature, bug fix, other change, dev setup

### CODE REVIEW

**Comment by jaheba** [comment]
Maybe r2c for row-2-col is a better name?

File: src/gluonts/itertools.py [hunk 1]
```diff
@@ -129,18 +149,79 @@ def __len__(self) -> int:
...
+    def next(self):
+        self.current = list(itertools.islice(self.iterable, self.length))
 
     def __iter__(self):
-        return itertools.islice(self.iterable, self.length)
+        yield from self.current
+        self.next()
 
     def __len__(self) -> int:
-        return len(self.iterable)
```
Comment (lostella) [inline src/gluonts/itertools.py:146]: So this was wrong?

File: src/gluonts/itertools.py [hunk 2]
```diff
@@ -129,18 +149,79 @@ def __len__(self) -> int:
...
+        return len(self.current)
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def rows_to_columns(
```
Comment (lostella) [inline src/gluonts/itertools.py:189]: These two are very simple, but we should probably have a few test cases

**Review by lostella — CHANGES_REQUESTED** [review]

File: src/gluonts/itertools.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/itertools.py:146]: Right, I'm wondering if it makes even sense to have `__len__` here. Also, we probably lacked tests for it.

File: src/gluonts/itertools.py (hunk 2 above)
Comment (jaheba) [reply src/gluonts/itertools.py:189]: They have a simple test case in case of the doctest!

File: src/gluonts/itertools.py (hunk 1 above)
Comment (jaheba) [reply src/gluonts/itertools.py:146]: I'm not so sure either, I guess I can remove it and check what happens?

File: src/gluonts/itertools.py (hunk 2 above)
Comment (lostella) [reply src/gluonts/itertools.py:189]: Do those run? I cannot remember honestly :-)

File: src/gluonts/itertools.py (hunk 1 above)
Comment (lostella) [reply src/gluonts/itertools.py:146]: Yes, let's do that

File: src/gluonts/itertools.py [hunk 3]
```diff
@@ -129,18 +149,79 @@ def __len__(self) -> int:
...
-    over multiple times.
+    over multiple times:
+
+        >>> isl = IterableSlice([1, 2, 3, 4, 5], 3)
+        >>> list(isl)
+        [1, 2, 3]
+        >>> list(isl)
+        [4, 5]
+        >>> list(isl)
+        []
```
Comment (lostella) [inline src/gluonts/itertools.py:164]: I'm not sure this is correct, I would expect `isl` to materialize to the same list every time

Comment (lostella) [reply src/gluonts/itertools.py:164]: `IterableSlice(iter([1, 2, 3, 4, 5]), 3)` would give what the docstring says

**Review by lostella — APPROVED** [review]

**Pull request merged**